test/balancerd/mzcompose.py

```diff
@@ -7,19 +7,56 @@
 # the Business Source License, use of this software will be governed
 # by the Apache License, Version 2.0.
 
+import json
+
 from pg8000.dbapi import ProgrammingError
 from pg8000.exceptions import InterfaceError
 
 from materialize.mzcompose.composition import Composition
 from materialize.mzcompose.services.balancerd import Balancerd
+from materialize.mzcompose.services.frontegg import Frontegg
 from materialize.mzcompose.services.materialized import Materialized
+from materialize.mzcompose.services.test_certs import TestCerts
+
+TENANT_ID = "00000000-0000-0000-0000-000000000000"
+ENCODING_KEY = "blah"
+USERS = json.dumps({"u1": {"client": "cl_id", "password": "pw"}})
+ADMIN_ROLE = "admin"
+ROLES = json.dumps({"u1": ["users", ADMIN_ROLE]})
+FRONTEGG_URL = "http://frontegg:6800/"
 
 SERVICES = [
-    Balancerd(),
+    TestCerts(),  # TODO: Is this right? Do the Balancerd and Frontegg services need to do something so they can get the test cert data?
+    Balancerd(
+        entrypoint=[
+            "balancerd",
+            "service",
+            "--pgwire-listen-addr=0.0.0.0:6875",
+            "--https-listen-addr=0.0.0.0:6876",
+            "--internal-http-listen-addr=0.0.0.0:6878",
+            "--frontegg-resolver-template=materialized:6875",
+            "--frontegg-jwk={TODO}",
+            f"--frontegg-api-token-url={FRONTEGG_URL}",
+            f"--frontegg-admin-role={ADMIN_ROLE}",
+            "--https-resolver-template='materialized:6876'",
+        ]
+    ),
+    Frontegg(
+        tenant_id=TENANT_ID,
+        encoding_key=ENCODING_KEY,
+        users=USERS,
+        roles=ROLES,
+    ),
     Materialized(
+        options=[
+            f"--frontegg-tenant-id={TENANT_ID}",
+            f"--frontegg-jwk={TODO}",
+            f"--frontegg-api_token_url={FRONTEGG_URL}",
+            f"--frontegg-admin_role={ADMIN_ROLE}",
+        ],
         # We do not do anything interesting on the Mz side
         # to justify the extra restarts
-        sanity_restart=False
+        sanity_restart=False,
     ),
```

Comment:
Thanks a lot! Can you add a couple more workflows to this mzcompose.py that check that authentication succeeds/fails with the right/wrong username and the right/wrong password?

Reply:
I will, but I need some help knowing how to structure this file. How can I get the `TestCerts` service and then plumb those certs down into the balancerd, frontegg, and mz services?

Reply:
I think this is cleared up? Otherwise see for example TestCerts in test/platform-checks/mzcompose.py and the secrets volume.